File: leetcode/52-n-queens-ii/main.go
package main

import "fmt"

/*
	 1st approach: backtracking
	 - - https://www.youtube.com/watch?v=5v6zdfkImms
	 - basically try every possisbilities within the safe region
	 - for each coordinate, we need to check the whole board to see if it is safe to place a queen

   Time    O(n^4) for each coordinate, we need to check if safe
   Space   O(n^2)
   4 ms, faster than 31.37%
*/
func totalNQueens(n int) int {
	// init board
	board := BoardConstructor(n)
	// result
	result := make(map[string]bool)

	var backtracking func(b *Board, col, n int)
	backtracking = func(b *Board, col, n int) {
		if col == n {
			// fmt.Println(b.M)
			result[b.Stringify()] = true
			return
		}
		for i := 0; i < n; i++ {
			if b.IsSafe(i, col) {
				b.Place(i, col)
				backtracking(b, col+1, n)
				b.Remove(i, col)
			}
		}
	}

	backtracking(&board, 0, n)
	return len(result)
}

// board class
type Board struct {
	M []string
	N int
}

func BoardConstructor(n int) Board {
	m := []string{}
	for i := 0; i < n; i++ {
		temp := ""
		for j := 0; j < n; j++ {
			temp += "."
		}
		m = append(m, temp)
	}
	return Board{m, n}
}

func (b *Board) Place(row, col int) {
	// b.M[row][col] = "Q"
	b.M[row] = b.M[row][:col] + "Q" + b.M[row][col+1:]
}

func (b *Board) Remove(row, col int) {
	// b.M[row][col] = "."
	b.M[row] = b.M[row][:col] + "." + b.M[row][col+1:]
}

func (b *Board) IsSafe(row, col int) bool {
	// check row and col
	for i := 0; i < b.N; i++ {
		if b.M[i][col] == 'Q' {
			return false
		}
		if b.M[row][i] == 'Q' {
			return false
		}
	}
	// check diagonal
	for i := 0; i < b.N; i++ {
		for j := 0; j < b.N; j++ {
			if i+j == row+col || i-j == row-col {
				if i != row && j != col && b.M[i][j] == 'Q' {
					return false
				}
			}
		}
	}
	return true
}

func (b *Board) Stringify() string {
	m := ""
	for i := 0; i < b.N; i++ {
		m += b.M[i]
	}
	return m
}

func main() {
	fmt.Println(totalNQueens(4))
	fmt.Println(totalNQueens(5))
	fmt.Println(totalNQueens(8))
}

File: leetcode/52-n-queens-ii/main.py
"""
    1st approach: backtracking
    - - similar to lc37, lc51
    - https://www.youtube.com/watch?v=5v6zdfkImms
    - basically try every possisbilities within the safe region
    - for each coordinate, we need to check the whole board to see if it is safe to place a queen

    Time    O(n^4) for each coordinate, we need to check if safe
    Space   O(n^2)
    340 ms, faster than 5.46%
"""


class Solution(object):

    def __init__(self):
        self.result = set()

    def totalNQueens(self, n):
        """
        :type n: int
        :rtype: List[List[str]]
        """
        b = Board(n)
        self.backtracking(b, 0, n)
        return len(self.result)

    def backtracking(self, b, row, n):
        if row == n:
            self.result.add(b.stringify())
            return
        for i in range(n):
            if b.isSafe(row, i):
                b.place(row, i)
                self.backtracking(b, row+1, n)
                b.remove(row, i)


class Board(object):
    def __init__(self, n):
        temp = []
        for i in range(n):
            temp.append(n*".")
        self.m = temp
        self.n = n

    def place(self, row, col):
        # basically it is self.m[row][col] = "Q"
        self.m[row] = self.m[row][:col]+"Q"+self.m[row][col+1:]

    def remove(self, row, col):
        # basically it is self.m[row][col] = "."
        self.m[row] = self.m[row][:col]+"."+self.m[row][col+1:]

    def isSafe(self, row, col):
        # check row and col O(n)
        for i in range(self.n):
            if self.m[i][col] == "Q":
                return False
            if self.m[row][i] == "Q":
                return False
        # check diagonal O(n^2)
        for i in range(self.n):
            for j in range(self.n):
                if i+j == row+col or i-j == row-col:
                    if i != row and j != col and self.m[i][j] == "Q":
                        return False
        return True

    def stringify(self):
        # O(n)
        temp = ""
        for i in range(self.n):
            temp += self.m[i]
        return temp


print(Solution().totalNQueens(4))
print(Solution().totalNQueens(5))
print(Solution().totalNQueens(8))
